lib/matplotlib/backends/registry.py

```diff
@@ -82,11 +280,130 @@ def list_builtin(self, filter_=None):
             Backend names.
         """
         if filter_ == BackendFilter.INTERACTIVE:
-            return self._BUILTIN_INTERACTIVE
+            return [k for k, v in self._BUILTIN_BACKEND_TO_GUI_FRAMEWORK.items()
+                    if v != "headless"]
         elif filter_ == BackendFilter.NON_INTERACTIVE:
-            return self._BUILTIN_NOT_INTERACTIVE
+            return [k for k, v in self._BUILTIN_BACKEND_TO_GUI_FRAMEWORK.items()
+                    if v == "headless"]
+
+        return [*self._BUILTIN_BACKEND_TO_GUI_FRAMEWORK]
+
+    def list_gui_frameworks(self):
+        """
+        Return list of GUI frameworks used by Matplotlib backends.
+
+        Returns
+        -------
+        list of str
+            GUI framework names.
+        """
+        return [k for k in self._GUI_FRAMEWORK_TO_BACKEND if k != "headless"]
+
+    def load_backend_module(self, backend):
+        """
+        Load and return the module containing the specified backend.
+
+        Parameters
+        ----------
+        backend : str
+            Name of backend to load.
+
+        Returns
+        -------
+        Module
+            Module containing backend.
+        """
+        module_name = self._backend_module_name(backend)
+        return importlib.import_module(module_name)
+
+    def resolve_backend(self, backend):
+        """
+        Return the backend and GUI framework for the specified backend name.
+
+        If the GUI framework is not yet known then it will be determined by loading the
+        backend module and checking the ``FigureCanvas.required_interactive_framework``
+        attribute.
+
+        This function only loads entry points if they have not already been loaded and
+        the backend is not built-in and not of ``module://some.backend`` format.
+
+        Parameters
+        ----------
+        backend : str or None
+            Name of backend, or None to use the default backend.
+
+        Returns
+        -------
+        Tuple of backend (str) and GUI framework (str or None).
+            A non-interactive backend returns None for its GUI framework rather than
+            "headless".
```

Comment:
Do we need to specify "rather than headless" here? This applies to both this and the following method.

Trying to see where a user of only public API would see "headless" at all, and thus adding "there is this other thing that might have been included" is only clouding what is happening.


If I'm missing where "headless" would be publicly visible, I think this is fine.

"headless" is a functional _input_ for `backend_for_gui_framework` (returning `"agg"`). However this is not explicitly stated in at least docstrings.

Further,

```python
>>> reg.resolve_gui_or_backend("headless")
('agg', 'headless')
```

Which contradicts the statement of "we return None for non-interactive"

I'd be happy to call anything to do with "headless" an internal implementation detail, just think it should probably then not leak to the docstring.